cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,273 @@
+{
+  auto output = make_numeric_column(
+    data_type(type_to_id<size_type>()), input.size(), mask_state::UNALLOCATED, stream, mr);
+  auto output_mutable_view = output->mutable_view();
+
+  // These device column views are necessary for creating iterators that work
+  // for columns of compound types. The column_view iterators fail for compound
+  // types because they return raw pointers to the start of the data.
+  auto input_device_view       = column_device_view::create(input, stream);
+  auto left_edges_device_view  = column_device_view::create(left_edges, stream);
+  auto right_edges_device_view = column_device_view::create(right_edges, stream);
+
+  // Compute the maximum shift required for either edge, then shift all the iterators appropriately.
+  size_type null_shift = max(left_edges.null_count(), right_edges.null_count());
+  auto left_begin      = left_edges_device_view->begin<T>();
+  auto left_end        = left_edges_device_view->end<T>();
+  auto right_begin     = right_edges_device_view->begin<T>();
+
+  if (edge_null_precedence == null_order::BEFORE) {
+    left_begin  = thrust::next(left_begin, null_shift);
+    right_begin = thrust::next(right_begin, null_shift);
+  } else {
+    left_end = thrust::prev(left_end, null_shift);
+  }
+
+  // If all the nulls are at the beginning, the indices found by lower_bound
+  // will be off by null_shift, but if they're at the end the indices will
+  // already be correct.
+  const size_type index_shift = (edge_null_precedence == null_order::BEFORE) ? null_shift : 0;
+
+  using RandomAccessIterator = decltype(left_edges_device_view->begin<T>());
+
+  if (input.has_nulls()) {
+    thrust::transform(rmm::exec_policy(stream),
+                      input_device_view->pair_begin<T, true>(),
+                      input_device_view->pair_end<T, true>(),
+                      output_mutable_view.begin<size_type>(),
+                      bin_finder<T, RandomAccessIterator, LeftComparator, RightComparator>(
+                        left_begin, left_end, right_begin, index_shift));
+  } else {
+    thrust::transform(rmm::exec_policy(stream),
+                      input_device_view->pair_begin<T, false>(),
+                      input_device_view->pair_end<T, false>(),
+                      output_mutable_view.begin<size_type>(),
+                      bin_finder<T, RandomAccessIterator, LeftComparator, RightComparator>(
+                        left_begin, left_end, right_begin, index_shift));
+  }
+
+  const auto mask_and_count = valid_if(output_mutable_view.begin<size_type>(),
+                                       output_mutable_view.end<size_type>(),
+                                       filter_null_sentinel());
+
+  output->set_null_mask(mask_and_count.first, mask_and_count.second);
+  return output;
+}
+
+template <typename T>
+constexpr auto is_supported_bin_type()
+{
+  return ((is_numeric<T>() && !std::is_same<T, bool>::value)) ||
+         std::is_same<T, string_view>::value;
```

Comment:
I wonder if this could be simply replaced with `cudf::is_relationally_comparable()` and `cudf::is_relationally_equal()` rather listing specific types.